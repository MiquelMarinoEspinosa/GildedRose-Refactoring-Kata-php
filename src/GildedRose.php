<?php

declare(strict_types=1);

namespace GildedRose;

final class GildedRose
{
    private const MINIMUM_ITEM_QUALITY = 0;
    private const MAXIMUM_ITEM_QUALITY = 50;
    private const AGED_BRIE_ITEM_NAME = 'Aged Brie';
    private const SULFURAS_ITEM_NAME = 'Sulfuras, Hand of Ragnaros';
    private const BACKSTAGE_PASSES_ITEM_NAME = 'Backstage passes to a TAFKAL80ETC concert';

    /**
     * @param Item[] $items
     */
    public function __construct(
        private array $items
    ) {
    }

    public function updateQuality(): void
    {
        foreach ($this->items as $item) {
            $this->updateItem($item);
        }
    }

    private function updateItem(Item $item): void
    {
        if ($item->name != self::AGED_BRIE_ITEM_NAME and $item->name != self::BACKSTAGE_PASSES_ITEM_NAME) {
            if ($item->quality > self::MINIMUM_ITEM_QUALITY) {
                if ($item->name != self::SULFURAS_ITEM_NAME) {
                    $item->quality = $item->quality - 1;
                }
            }
        } else {
            if ($item->quality < self::MAXIMUM_ITEM_QUALITY) {
                $this->increaseItemQuality($item);;
                if ($item->name == self::BACKSTAGE_PASSES_ITEM_NAME) {
                    if ($item->sellIn < 11) {
                        if ($item->quality < self::MAXIMUM_ITEM_QUALITY) {
                            $this->increaseItemQuality($item);;
                        }
                    }
                    if ($item->sellIn < 6) {
                        if ($item->quality < self::MAXIMUM_ITEM_QUALITY) {
                            $this->increaseItemQuality($item);;
                        }
                    }
                }
            }
        }

        if ($item->name != self::SULFURAS_ITEM_NAME) {
            $item->sellIn = $item->sellIn - 1;
        }

        if ($item->sellIn < 0) {
            if ($item->name != self::AGED_BRIE_ITEM_NAME) {
                if ($item->name != self::BACKSTAGE_PASSES_ITEM_NAME) {
                    if ($item->quality > self::MINIMUM_ITEM_QUALITY) {
                        if ($item->name != self::SULFURAS_ITEM_NAME) {
                            $item->quality = $item->quality - 1;
                        }
                    }
                } else {
                    $item->quality = self::MINIMUM_ITEM_QUALITY;
                }
            } else {
                if ($item->quality < self::MAXIMUM_ITEM_QUALITY) {
                    $this->increaseItemQuality($item);
                }
            }
        }
    }

    private function increaseItemQuality(Item $item): void
    {
        $item->quality = $item->quality + 1;
    }
}
