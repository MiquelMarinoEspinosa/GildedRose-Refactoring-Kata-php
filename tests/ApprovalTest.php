<?php

declare(strict_types=1);

namespace Tests;

use GildedRose\GildedRose;
use GildedRose\Item;
use PHPUnit\Framework\TestCase;
use ApprovalTests\Approvals;

/**
 * This unit test uses [Approvals](https://github.com/approvals/ApprovalTests.php).
 *
 * There are two test cases here with different styles:
 * <li>"foo" is more similar to the unit test from the 'Java' version
 * <li>"thirtyDays" is more similar to the TextTest from the 'Java' version
 *
 * I suggest choosing one style to develop and deleting the other.
 */
class ApprovalTest extends TestCase
{

    public function testListItems(): void
    {
        $items = [
            new Item('foo', 0, 0),
            new Item('foo', 0, 1),
            new Item('Aged Brie', 0, 1),
            new Item('Backstage passes to a TAFKAL80ETC concert', 0, 1),
            new Item('foo', 0, 2),
            new Item('Aged Brie', 0, 50),
            new Item('Aged Brie', 11, 48),
            new Item('Backstage passes to a TAFKAL80ETC concert', 10, 48),
            new Item('Backstage passes to a TAFKAL80ETC concert', 5, 48),
        ];
        $app = new GildedRose($items);
        $app->updateQuality();

        Approvals::verifyList($items);
    }

    public function testThirtyDays(): void
    {
        ob_start();

        $argv = ["", "30"];
        include(__DIR__ . '/../fixtures/texttest_fixture.php');

        $output = ob_get_clean();

        Approvals::verifyString($output);
    }
}
