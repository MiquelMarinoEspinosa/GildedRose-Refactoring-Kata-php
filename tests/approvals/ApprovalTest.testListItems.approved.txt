[0] -> foo, -1, 0
[1] -> foo, -1, 0
[2] -> Aged Brie, -1, 3
[3] -> Backstage passes to a TAFKAL80ETC concert, -1, 0
[4] -> foo, -1, 0
[5] -> Aged Brie, -1, 50
[6] -> Aged Brie, 10, 49
[7] -> Backstage passes to a TAFKAL80ETC concert, 9, 50
[8] -> Backstage passes to a TAFKAL80ETC concert, 4, 50
